### PR TITLE
feat(ui): add TPS metric to token usage display

### LINKED ISSUE #72: Add TPS (Tokens Per Second) metric to token usage display
Add a derived TPS (tokens per second) metric to the `TokenUsageDisplay` component to show the token generation speed during LLM responses. This metric provides useful insight into model performance and helps users understand response latency characteristics.

### Background

The existing token usage metrics already track:
- `completion_tokens` — number of tokens generated
- `ttft` (Time to First Token) — latency until first token arrives
- `ttr` (Time to Response) — total response time

TPS can be calculated as a derived metric: `completion_tokens / ((ttr - ttft) / 1000)`, which measures actual token generation speed excluding the initial model processing time.

### Scope

- **UI (`ui/src/components/chat/TokenUsageDisplay.tsx`):**
  - Add a `useMemo` hook to calculate TPS from existing `completion_tokens`, `ttr`, and `ttft` values.
  - Formula: `completion_tokens / ((ttr - ttft) / 1000)` tokens per second.
  - Handle edge cases: return `null` if any required values are missing or if `(ttr - ttft) <= 0`.
  - Add a new metric display item with:
    - Icon: `MdSpeed` from `react-icons/md`
    - Format: `{tps} t/s` (e.g., "42.5 t/s")
    - Tooltip: "Generation Speed: {tps} tokens/sec"
  - Only display the TPS metric when a valid positive value can be calculated.
  - Position alongside existing metrics (after TTR or before token counts).
- **UI (`ui/src/types/chat.ts`):**
  - No changes needed — TPS is derived, not stored.

### Acceptance criteria

- TPS is displayed in the token usage bar when `completion_tokens`, `ttr`, and `ttft` are all available and valid.
- TPS is hidden when any required value is missing or when the calculation would be invalid (e.g., `ttr <= ttft`).
- The displayed value is formatted to one decimal place (e.g., "42.5 t/s").
- Tooltip provides clear explanation of what the metric represents.
- Visual styling matches existing metrics (icon size, spacing, hover states).
- No regressions to existing token usage display functionality.

### How to test manually

1. Run the Engine and UI locally (see `engine/README.md` and `ui/README.md`).
2. Start a conversation and send a message.
3. Wait for the agent to respond and observe the token usage bar below the response.
4. Verify that a TPS metric appears (e.g., "42.5 t/s") alongside TTFT, TTR, and token counts.
5. Hover over the TPS metric and confirm the tooltip shows "Generation Speed: X tokens/sec".
6. Test with a very short response to verify TPS handles edge cases gracefully.
7. Verify TPS does not appear if `ttft` or `ttr` are missing (e.g., for cached/historical messages without timing data).

### PR DESCRIPTION
**Pull request by Mahadevan2005**
## Summary
Adds a TPS (tokens per second) metric to the `TokenUsageDisplay` component to show token generation speed during LLM responses.

## Related Issue
Closes #72

## Changes Made
- Added `useMemo` hook to calculate TPS from `completion_tokens`, `ttr`, and `ttft`
- Formula: `completion_tokens / ((ttr - ttft) / 1000)` tokens per second
- Added TPS display with `MdSpeed` icon formatted as "X t/s"
- Added tooltip: "Generation Speed: X tokens/sec"
- Used `hasTPS` guard for consistent render pattern (matches `hasTokenUsage`, `hasTTFT`, `hasTTR`)
- Positioned after TTR metric in the usage bar
- Handles edge cases: null values, invalid calculations, division by zero

## Type of Change
- [x] New feature (non-breaking change which adds functionality)

## Testing
- [x] TypeScript compilation passes (`npm run build`)
- [x] Linting passes - no new errors introduced
- [x] Build succeeds in production mode
- [x] Code follows existing patterns and conventions
- [x] Follows Skyflo.ai coding standards (TypeScript, Airbnb style)

## Verification
```
✓ Compiled successfully
✓ Checking validity of types
✓ No linting errors in TokenUsageDisplay.tsx
✓ Build completed successfully
```

## Notes
- No new linting errors introduced - TokenUsageDisplay.tsx passes all checks
- Pre-existing linting errors in other files remain unchanged
- Implementation follows existing component patterns (`has*` naming convention)

### CODE REVIEW

**Comment by Mahadevan2005** [comment]
@KaranJagtiani the changes are done! 

- Squashed all commits into one
- No lock file changes
- Only 2 code files modified

PR is now clean and ready.

File: ui/src/components/chat/ChatInterface.tsx
Comment (KaranJagtiani) [inline ui/src/components/chat/ChatInterface.tsx]: The `lastTTRUpdateRef` needs to be reset here as well.

```suggestion
        lastTTRUpdateRef.current = 0;
        requestStartTimeRef.current = Date.now();
```

**Comment by Mahadevan2005** [comment]
@KaranJagtiani I've implemented issue #75! Here's what was done:

### Backend Changes:
- Added `duration_ms` (int) to completed SSE event payload in `graph.py`
- Updated `record_ttr()` in `agent.py` to prefer `duration_ms` over `duration * 1000`
- Maintained backwards compatibility with existing `duration` field

### Frontend Changes:
- Added `duration_ms?: number` to `CompletedEvent` type
- Updated `onComplete` callback to accept optional `duration_ms` parameter  
- Parse and propagate server duration from SSE events
- Prefer server-measured TTR over client calculation
- Keep client timing as fallback for older engines
- Added missing `lastTTRUpdateRef` resets

### Result:
- TPS now uses accurate server timing (no network latency)
- TTR remains consistent when reloading conversations
- No breaking changes to SSE contract
- Fully backwards compatible

File: ui/src/components/chat/ChatInterface.tsx [hunk 1]
```diff
@@ -457,9 +459,19 @@ export function ChatInterface({ conversationId }: ChatInterfaceProps) {
         approvalDecisionRef.current = null;
       },
 
-      onComplete: () => {
-        if (requestStartTimeRef.current) {
-          const ttr = Date.now() - requestStartTimeRef.current;
+      onComplete: (duration_ms?: number) => {
+        // Prefer server-measured duration_ms, fallback to client timing
```
Comment (KaranJagtiani) [inline ui/src/components/chat/ChatInterface.tsx]: Remove redundant comments.

File: engine/src/api/agent/graph.py
Comment (KaranJagtiani) [inline engine/src/api/agent/graph.py]: We need to emit the `duration` / `duration_ms` for stopped runs too.

```suggestion
            except StopRequested:
                end_time = time.time()
                start_time = get_state_value(initial_state, "start_time", end_time)
                duration = end_time - start_time
                duration_ms = int(duration * 1000)

                if self.event_callback:
                    await self.event_callback(
                        {
                            "type": "completed",
                            "status": "stopped",
                            "run_id": get_state_value(initial_state, "run_id"),
                            "duration": duration,
                            "duration_ms": duration_ms,
                        }
                    )
```

File: ui/src/components/chat/ChatInterface.tsx
Comment (KaranJagtiani) [inline ui/src/components/chat/ChatInterface.tsx]: The same block never resets requestStartTimeRef or lastTTRUpdateRef before firing the next approval stream. Every approval after the first therefore measures elapsed time starting from the very first tool. While the server now emits duration_ms at completion, all streamed TPS updates (and any fallback path when duration_ms is absent) are computed off this stale timestamp, so the denominator grows across the entire batch and misstates generation speed. Reset the start time and throttle ref before invoking startApprovalStream.

**Comment by Mahadevan2005** [comment]
@KaranJagtiani All bulk approval and aggregate TPS issues fixed!

**Changes Made:**

### Bulk Approval Fixes:
- Added `resetLiveUsage()` before each approval in bulk continuation
- Added `lastTTRUpdateRef.current = 0` before each approval
- Added `requestStartTimeRef.current = Date.now()` before each approval
- Prevents metric accumulation across sequential approvals

### Aggregate TPS Fixes:
- Added `total_generation_ms` to TokenUsage type
- Accumulate generation time in `accumulateUsage()` function
- Updated TPS calculation to prefer `total_generation_ms` for aggregates
- Keep per-message TPS using `(ttr - ttft)` for individual displays

**Testing Logic:**

**Single Turn:**
- Uses `(ttr - ttft)` → Per-message TPS

**Multi-Turn Footer:**
- Turn 1: 10 tokens, 1000ms generation
- Turn 2: 20 tokens, 1000ms generation
- Turn 3: 15 tokens, 1000ms generation
- Total: 45 tokens / 3000ms = 15 t/s

**Bulk Approvals:**
- Approval 1: Fresh metrics from 0
- Approval 2: Fresh metrics from 0 (resets before starting)
- Approval 3: Fresh metrics from 0 (resets before starting)

TPS feature is now ready for all workflows.

File: ui/src/components/chat/ChatInterface.tsx [hunk 2]
```diff
@@ -594,14 +618,29 @@ export function ChatInterface({ conversationId }: ChatInterfaceProps) {
         if (source !== "main") {
           return;
         }
+        const now = Date.now();
+        const elapsedSinceRequestStart =
+          requestStartTimeRef.current != null
+            ? now - requestStartTimeRef.current
+            : undefined;
+
+        // Throttle TTR updates so we don't re-render too often while streaming
```
Comment (KaranJagtiani) [inline ui/src/components/chat/ChatInterface.tsx]:
```suggestion
        // Throttle TTR updates to prevent excessive re-renders during streaming
```

**Comment by Mahadevan2005** [comment]
@KaranJagtiani Done! 

- Updated comment wording as requested
- Squashed all commits into one

Ready to merge!

**Review by KaranJagtiani — APPROVED** [review]
Great work @Mahadevan2005 👏 Closing two issues in a single PR is well done.

**Comment by Mahadevan2005** [comment]
@KaranJagtiani Thank you so much for the thorough reviews and guidance throughout this PR! 🙏

This was an incredible learning experience - from implementing the initial TPS metric to expanding the scope to fix the server-measured TTR architecture. Your detailed feedback helped me understand not just what to change, but why it matters.

Special thanks for:
- Catching the bulk approval metric accumulation issue
- Guiding the aggregate TPS calculation fix
- The patience through 15+ review cycles
- Teaching me about backwards compatibility and proper API design

I learned a ton about, real-time systems, and professional code review processes. Looking forward to contributing more to Skyflo.ai! 🚀

Thanks again for the mentorship!
